### ISSUE TITLE
qtile should not use LD_LIBRARY_PATH

### ISSUE THREAD

**Issue by avnik** [NixOS/nixpkgs#11219]
```
[avn@bulldozer:pts/18]/home/avn/invoicing% zathura example.pdf                                                                                                     
** (zathura:17204): WARNING **: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
/nix/store/75b1vkjfia3x6qwn7j592ypg33xj1vjh-zathura-core-0.3.3/bin/zathura: symbol lookup error: /nix/store/rwcwfdghyxxl9jzznly23y29iaz438i3-gtk+3-3.18.4/lib/libgtk-3.so.0: undefined symbol: g_param_spec_get_name_quark
```
Looks like libgtk3 miss 1 symbol from glib and 6 from pango

```
[avn@bulldozer:pts/18]~% ldd -r /nix/store/rwcwfdghyxxl9jzznly23y29iaz438i3-gtk+3-3.18.4/lib/libgtk-3.so.0.1800.4                                                                            ~
	linux-vdso.so.1 (0x00007ffdbd9fe000)
	libgdk-3.so.0 => /nix/store/rwcwfdghyxxl9jzznly23y29iaz438i3-gtk+3-3.18.4/lib/libgdk-3.so.0 (0x00007fc77918c000)
	libXinerama.so.1 => /nix/store/56ps0jrhf0q7d7mdhgiwqk2pggf41dzn-libXinerama-1.1.3/lib/libXinerama.so.1 (0x00007fc778f84000)
	libXrandr.so.2 => /nix/store/9nmm2i2rlhl4x6rmzv634zy9c2g53ssa-libXrandr-1.5.0/lib/libXrandr.so.2 (0x00007fc778d74000)
	libXcursor.so.1 => /nix/store/ggwf1p2dkiab59v7j9sdrns7c5yrh4hq-libXcursor-1.1.14/lib/libXcursor.so.1 (0x00007fc778b64000)
	libpangocairo-1.0.so.0 => /nix/store/1hgr2q3mfixs10kj9yzqyyrqbyzjkm29-pango-1.36.8/lib/libpangocairo-1.0.so.0 (0x00007fc778954000)
	libXi.so.6 => /nix/store/x75lb2x9anzka6qwn8f0zi8bc33djnzs-libXi-1.7.5/lib/libXi.so.6 (0x00007fc778744000)
	libXcomposite.so.1 => /nix/store/qw1a6xzkb35afsg317dm209iwvk9dj5v-libXcomposite-0.4.4/lib/libXcomposite.so.1 (0x00007fc77853c000)
	libcairo-gobject.so.2 => /nix/store/fl7srm0d8z5zdm9gp2diqr67rkfsxd36-cairo-1.14.2/lib/libcairo-gobject.so.2 (0x00007fc778334000)
	libcairo.so.2 => /nix/store/fl7srm0d8z5zdm9gp2diqr67rkfsxd36-cairo-1.14.2/lib/libcairo.so.2 (0x00007fc77800c000)
	libpixman-1.so.0 => /nix/store/100rzjkjc3293h4pbwcqkc58f49va3mz-pixman-0.32.8/lib/libpixman-1.so.0 (0x00007fc777d5c000)
	libEGL.so.1 => /run/opengl-driver/lib/libEGL.so.1 (0x00007fc777b2c000)
	libgbm.so.1 => /run/opengl-driver/lib/libgbm.so.1 (0x00007fc77791c000)
	libwayland-server.so.0 => /nix/store/1964rliyxykqpg57f5gvlg6g6klm5lvq-wayland-1.9.0/lib/libwayland-server.so.0 (0x00007fc777704000)
	libxcb-shm.so.0 => /nix/store/r3xv21sfbky2pi0x8gypidal63m057f7-libxcb-1.11.1/lib/libxcb-shm.so.0 (0x00007fc7774fc000)
	libXrender.so.1 => /nix/store/daj5y488f74q24jgxr4i2gj1ri58970a-libXrender-0.9.9/lib/libXrender.so.1 (0x00007fc7772ec000)
	libGL.so.1 => /run/opengl-driver/lib/libGL.so.1 (0x00007fc77705c000)
	libglapi.so.0 => /run/opengl-driver/lib/libglapi.so.0 (0x00007fc776e2c000)
	libXdamage.so.1 => /nix/store/20x2pzahzk7nz9lndvzhb37cypvzgf3n-libXdamage-1.1.4/lib/libXdamage.so.1 (0x00007fc776c24000)
	libXfixes.so.3 => /nix/store/4jpalpbkxc6kfz6filfcsq9x0jqqdxmp-libXfixes-5.0.1/lib/libXfixes.so.3 (0x00007fc776a1c000)
	libX11-xcb.so.1 => /nix/store/frg4x2wa979bvw6ahac5wvi0bx3kf53w-libX11-1.6.3/lib/libX11-xcb.so.1 (0x00007fc776814000)
	libxcb-glx.so.0 => /nix/store/r3xv21sfbky2pi0x8gypidal63m057f7-libxcb-1.11.1/lib/libxcb-glx.so.0 (0x00007fc7765fc000)
	libxcb-dri2.so.0 => /nix/store/r3xv21sfbky2pi0x8gypidal63m057f7-libxcb-1.11.1/lib/libxcb-dri2.so.0 (0x00007fc7763f4000)
	libxcb-dri3.so.0 => /nix/store/r3xv21sfbky2pi0x8gypidal63m057f7-libxcb-1.11.1/lib/libxcb-dri3.so.0 (0x00007fc7761ec000)
	libxcb-present.so.0 => /nix/store/r3xv21sfbky2pi0x8gypidal63m057f7-libxcb-1.11.1/lib/libxcb-present.so.0 (0x00007fc775fe4000)
	libxcb-randr.so.0 => /nix/store/r3xv21sfbky2pi0x8gypidal63m057f7-libxcb-1.11.1/lib/libxcb-randr.so.0 (0x00007fc775dd4000)
	libxcb-xfixes.so.0 => /nix/store/r3xv21sfbky2pi0x8gypidal63m057f7-libxcb-1.11.1/lib/libxcb-xfixes.so.0 (0x00007fc775bcc000)
	libxcb-render.so.0 => /nix/store/r3xv21sfbky2pi0x8gypidal63m057f7-libxcb-1.11.1/lib/libxcb-render.so.0 (0x00007fc7759bc000)
	libxcb-shape.so.0 => /nix/store/r3xv21sfbky2pi0x8gypidal63m057f7-libxcb-1.11.1/lib/libxcb-shape.so.0 (0x00007fc7757b4000)
	libxcb-sync.so.1 => /nix/store/r3xv21sfbky2pi0x8gypidal63m057f7-libxcb-1.11.1/lib/libxcb-sync.so.1 (0x00007fc7755ac000)
	libxshmfence.so.1 => /nix/store/ba4ma29s4dvvs6jas07nf9yjgfl9y5ig-libxshmfence-1.2/lib/libxshmfence.so.1 (0x00007fc7753a4000)
	libXxf86vm.so.1 => /nix/store/asc1fqbb5qqh8h1xyziz4gdp4l8yydbx-libXxf86vm-1.1.4/lib/libXxf86vm.so.1 (0x00007fc77519c000)
	libXext.so.6 => /nix/store/nwa59yl3sz6ldrhg94g45pf9rn0j8nwr-libXext-1.3.3/lib/libXext.so.6 (0x00007fc774f84000)
	libxcb.so.1 => /nix/store/r3xv21sfbky2pi0x8gypidal63m057f7-libxcb-1.11.1/lib/libxcb.so.1 (0x00007fc774d64000)
	libXau.so.6 => /nix/store/r3zpj4sa3bn8jxr2zk8k2ax9y7mrkhfj-libXau-1.0.8/lib/libXau.so.6 (0x00007fc774b5c000)
	libXdmcp.so.6 => /nix/store/n3ybgbqy68pr7810j95xq79dmj2687cb-libXdmcp-1.1.2/lib/libXdmcp.so.6 (0x00007fc774954000)
	libdrm.so.2 => /nix/store/z9l0w0dpqflmfkyjx6wyi3718xh1s68q-libdrm-2.4.65/lib/libdrm.so.2 (0x00007fc774744000)
	libgdk_pixbuf-2.0.so.0 => /nix/store/yxdzfjqmc63gxq1kwk6zg2h2xhscmkrc-gdk-pixbuf-2.32.1/lib/libgdk_pixbuf-2.0.so.0 (0x00007fc77451c000)
	libatk-bridge-2.0.so.0 => /nix/store/fhrwdhjbj5cdzyzg997jippkbh76xvrv-at-spi2-atk-2.18.0/lib/libatk-bridge-2.0.so.0 (0x00007fc7742ec000)
	libatk-1.0.so.0 => /nix/store/ggdlrcjk4q4h1vf0vvdlrpv4blhj1k2k-atk-2.18.0/lib/libatk-1.0.so.0 (0x00007fc7740c4000)
	libatspi.so.0 => /nix/store/qcqd6fhp20rb4hdipjgzcyrli95i9qlk-at-spi2-core-2.18.0/lib/libatspi.so.0 (0x00007fc773e94000)
	libSM.so.6 => /nix/store/v8vvnzib0ijfdjxx1i83qixh5arxlp3j-libSM-1.2.2/lib/libSM.so.6 (0x00007fc773c8c000)
	libICE.so.6 => /nix/store/smnk65pdkc99imchbfadpfb4f0j6gbv6-libICE-1.0.9/lib/libICE.so.6 (0x00007fc773a6c000)
	libX11.so.6 => /nix/store/frg4x2wa979bvw6ahac5wvi0bx3kf53w-libX11-1.6.3/lib/libX11.so.6 (0x00007fc77372c000)
	libdbus-1.so.3 => /nix/store/0gwqsj2xbs58dia6aklhbaqz6dzf3c98-dbus-libs-1.8.20/lib/libdbus-1.so.3 (0x00007fc7734e4000)
	libxkbcommon.so.0 => /nix/store/ibm01l5fzncap0ablkvxgjmgx68sq49q-libxkbcommon-0.5.0/lib/libxkbcommon.so.0 (0x00007fc7732a4000)
	libwayland-cursor.so.0 => /nix/store/1964rliyxykqpg57f5gvlg6g6klm5lvq-wayland-1.9.0/lib/libwayland-cursor.so.0 (0x00007fc77309c000)
	libwayland-egl.so.1 => /run/opengl-driver/lib/libwayland-egl.so.1 (0x00007fc772e94000)
	libwayland-client.so.0 => /nix/store/1964rliyxykqpg57f5gvlg6g6klm5lvq-wayland-1.9.0/lib/libwayland-client.so.0 (0x00007fc772c84000)
	librt.so.1 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/librt.so.1 (0x00007fc772a7c000)
	libepoxy.so.0 => /nix/store/0hl15p8v2lz3cpqwkry910cm6w0lzh9q-epoxy-1.3.1/lib/libepoxy.so.0 (0x00007fc772784000)
	libpangoft2-1.0.so.0 => /nix/store/1hgr2q3mfixs10kj9yzqyyrqbyzjkm29-pango-1.36.8/lib/libpangoft2-1.0.so.0 (0x00007fc77256c000)
	libharfbuzz.so.0 => /nix/store/im0m6nyg2vmpw3czghymhwc9h4xx1m1i-harfbuzz-1.0.6/lib/libharfbuzz.so.0 (0x00007fc772304000)
	libgraphite2.so.3 => /nix/store/n3c5392g317zik4bb7fj6lj0af4xb01g-graphite2-1.2.4/lib/libgraphite2.so.3 (0x00007fc7720dc000)
	libpango-1.0.so.0 => /nix/store/1hgr2q3mfixs10kj9yzqyyrqbyzjkm29-pango-1.36.8/lib/libpango-1.0.so.0 (0x00007fc771e8c000)
	libgthread-2.0.so.0 => /nix/store/zy7xi64lsmnx0i0d0h8qgqigs4m85crc-glib-2.44.1/lib/libgthread-2.0.so.0 (0x00007fc771c84000)
	libfontconfig.so.1 => /nix/store/2msjjbmmrcx4cjqr76v31l8kvkrqkcq9-fontconfig-2.11.1/lib/libfontconfig.so.1 (0x00007fc771a44000)
	libexpat.so.1 => /nix/store/8hvsmq24ap9mnvx3x96gmq1qm86kyx8s-expat-2.1.0/lib/libexpat.so.1 (0x00007fc771814000)
	libfreetype.so.6 => /nix/store/v42khwmgdddlyqwmh6an87pvk9x4h8bw-freetype-2.5.4/lib/libfreetype.so.6 (0x00007fc77150c000)
	libbz2.so.1.0 => /nix/store/0iv2ykxshmq5gwfm8qbc4n7649algn3h-bzip2-1.0.6/lib/libbz2.so.1.0 (0x00007fc7712fc000)
	libpng16.so.16 => /nix/store/y7qsx1scsxxh9yvc1yjx9dx7vs13rljq-libpng-apng-1.6.18/lib/libpng16.so.16 (0x00007fc7710c4000)
	libgio-2.0.so.0 => /nix/store/zy7xi64lsmnx0i0d0h8qgqigs4m85crc-glib-2.44.1/lib/libgio-2.0.so.0 (0x00007fc770d44000)
	libgmodule-2.0.so.0 => /nix/store/zy7xi64lsmnx0i0d0h8qgqigs4m85crc-glib-2.44.1/lib/libgmodule-2.0.so.0 (0x00007fc770b3c000)
	libdl.so.2 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/libdl.so.2 (0x00007fc770934000)
	libz.so.1 => /nix/store/s6zky9ysv3qpwz53xvza43n8vzijyafn-zlib-1.2.8/lib/libz.so.1 (0x00007fc77071c000)
	libresolv.so.2 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/libresolv.so.2 (0x00007fc770504000)
	libgobject-2.0.so.0 => /nix/store/zy7xi64lsmnx0i0d0h8qgqigs4m85crc-glib-2.44.1/lib/libgobject-2.0.so.0 (0x00007fc7702ac000)
	libffi.so.6 => /nix/store/yp3n0vk0wggj1fm1bwagwssih9rarq1c-libffi-3.2.1/lib/../lib64/libffi.so.6 (0x00007fc77009c000)
	libglib-2.0.so.0 => /nix/store/zy7xi64lsmnx0i0d0h8qgqigs4m85crc-glib-2.44.1/lib/libglib-2.0.so.0 (0x00007fc76fd64000)
	libpthread.so.0 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/libpthread.so.0 (0x00007fc76fb44000)
	libm.so.6 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/libm.so.6 (0x00007fc76f83c000)
	libc.so.6 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/libc.so.6 (0x00007fc76f49c000)
	/nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib64/ld-linux-x86-64.so.2 (0x00005646621e1000)
	libuuid.so.1 => /nix/store/fr25hj7nh6hiwayzj7npwk2bb2h7zgr3-util-linux-2.27.1/lib/libuuid.so.1 (0x00007fc76f294000)
undefined symbol: pango_attr_foreground_alpha_new	(/nix/store/rwcwfdghyxxl9jzznly23y29iaz438i3-gtk+3-3.18.4/lib/libgtk-3.so.0.1800.4)
undefined symbol: g_param_spec_get_name_quark	(/nix/store/rwcwfdghyxxl9jzznly23y29iaz438i3-gtk+3-3.18.4/lib/libgtk-3.so.0.1800.4)
undefined symbol: pango_fc_font_map_config_changed	(/nix/store/rwcwfdghyxxl9jzznly23y29iaz438i3-gtk+3-3.18.4/lib/libgtk-3.so.0.1800.4)
undefined symbol: pango_renderer_set_alpha	(/nix/store/rwcwfdghyxxl9jzznly23y29iaz438i3-gtk+3-3.18.4/lib/libgtk-3.so.0.1800.4)
undefined symbol: pango_renderer_get_alpha	(/nix/store/rwcwfdghyxxl9jzznly23y29iaz438i3-gtk+3-3.18.4/lib/libgtk-3.so.0.1800.4)
undefined symbol: pango_attr_font_features_new	(/nix/store/rwcwfdghyxxl9jzznly23y29iaz438i3-gtk+3-3.18.4/lib/libgtk-3.so.0.1800.4)
undefined symbol: pango_attr_background_alpha_new	(/nix/store/rwcwfdghyxxl9jzznly23y29iaz438i3-gtk+3-3.18.4/lib/libgtk-3.so.0.1800.4)
```

**Comment by lethalman** [comment]
That's weird. What about `nix-store -qR /nix/store/75b1vkjfia3x6qwn7j592ypg33xj1vjh-zathura-core-0.3.3 | grep gtk` ? Either you have an `LD_LIBRARY_PATH` somewhere, or zathura uses dynamic loading and the gtk version used for it is wrong.

**Comment by avnik** [comment]
```
[avn@bulldozer:pts/18]~% nix-store -qR /nix/store/75b1vkjfia3x6qwn7j592ypg33xj1vjh-zathura-core-0.3.3 | grep gtk                                                                             ~
/nix/store/rwcwfdghyxxl9jzznly23y29iaz438i3-gtk+3-3.18.4
```

**Comment by lethalman** [comment]
Ok so it looks like zathura has been compiled with gtk 3.18 and pango 3.16, but gtk 3.18 closure has pango 3.18 instead. So, again, it must be an `LD_LIBRARY_PATH¿ somewhere.

**Comment by lethalman** [comment]
As per IRC, this is an `LD_LIBRARY_PATH` problem somewhere. If possible, please try to understand where it's coming from, so we may try to patch it. Closing this, feel free to open another one if you understand which nixos module is mining `LD_LIBRARY_PATH`.

**Comment by avnik** [comment]
qtile (windowmanager) wrapper was

**Comment by vcunat** [comment]
Has that been fixed? It isn't nice if a WM passes `$LD_LIBRARY_PATH` into env of all executables under it.

**Comment by lethalman** [comment]
@vcunat I think @avnik was working on this already, but this issue is not about zathura being broken anymore.

**Comment by lethalman** [comment]
Reopening with different title perhaps...

**Comment by kamilchm** [comment]
What's the simplest way to set ``LD_LIBRARY_PATH`` for buildPhase only?

**Comment by vcunat** [comment]
Just define it within the attrset passed to `stdenv.mkDerivation`.

**Comment by avnik** [comment]
@kamilchm  I discuss possible package improvements in irc, I sure, that these part of conversation may be interesting for you:
<pre>
11:34 < Lethalman> avn, so you have to understand where that LD_LIBRARY_PATH comes from, it might be a nixos module, in which case we may need to fix it to avoid using LD_LIBRARY_PATH
11:35 < avn> Lethalman: qtile (window-manager) wrapper export them
11:36 < Lethalman> I see
11:37 < Lethalman> avn, btw LD_LIBRARY_PATH= zathura should work
11:37 < Lethalman> avn, and someone better fix the qtile derivation to avoid LD_LIBRARY_PATH :)
11:40 < avn> Lethalman: thinking about how to fix... It accessing 
libxcb and pango via cffi
11:44 < avn> Lethalman: https://github.com/NixOS/nixpkgs/blob/mast
er/pkgs/applications/window-managers/qtile/default.nix#L31
11:44 < Lethalman> avn, not that, the qtile code...
11:46 < avn> Lethalman: https://github.com/qtile/qtile/blob/develo
p/libqtile/pangocffi.py#L61 loading itself
11:47 < Lethalman> avn, perfect, replace that string with an absol
ute path :)
11:48 < Lethalman> avn, avn, gobject = ffi.dlopen('${glib}/lib/lib
gobject-2.0.so.0')
11:51 < avn> Lethalman: btw, if you start look to qtile package --
 may be you hint me any idea about improving this: https://github.
com/NixOS/nixpkgs/blob/master/pkgs/applications/window-managers/qt
ile/restart_executable.patch
11:55 < avn> Lethalman: I thinking about  put  export QTILE_WRAPPER=$0 in wrapper, and use env["QTILE_WRAPPER"] for restart, so it should "catch" cases when /run/current.../bin/qtile updated, and I want to restart to new one
11:56 < Lethalman> avn, ah, that's a good idea yes
</pre>

@kamilchm any toughts about absolute paths, and restart modification?
(I can do this).

Another thing, which I don't know how to improve -- PATH and PYTHONPATH also have some stuff, related only to this qtile derivation (come from wrappers), but they taint any child process environment (like xterms)

**Comment by lethalman** [comment]
About PATH and PYTHONPATH, they should probably be unset when child processes are spawned by qtile.

**Comment by avnik** [comment]
PYTHONPATH can be just unset, but not PATH

btw they can be unset in first lines on unwrapped bin/qtile, as I know mangling os.environ["PYTHONPATH"] not affect current interpreter, only child processes

**Comment by kamilchm** [comment]
@avnik :+1: for absolute paths, I was thinking about it yesterday but I don't have time to do it now.
Restart modification looks good too.
I'll be happy with your patch.

**Comment by vcunat** [comment]
Note that unsetting `$LD_LIBRARY_PATH` will also cause problems to some programs.

**Comment by lethalman** [comment]
@vcunat I don't think anybody wanted to unset `LD_LIBRARY_PATH`. Of course also for PATH and PYTHONPATH, it's probably best to store somewhere the original env and reset in qtile.

**Comment by vcunat** [comment]
I saw the suggestion to run `LD_LIBRARY_PATH= zathura` (above), so I just wanted to warn it won't always work (well).

**Comment by kamilchm** [comment]
What is the best way to test it? I'll try to find some time to fix it.

**Comment by avnik** [comment]
Fixed in #11386

**Comment by avnik** [comment]
LD_LIBRARY_PATH now just not touched in wrapper (I change all paths to absolute),  PATH and PYTHONPATH saved in wrapper, and restored in python code
